docs/usecases/testing.md

```diff
@@ -72,3 +72,364 @@ Test aspects are powerful tools for modifying behavior of individual tests or ev
+more or less the same amount of time. Biggest source of complexity during testing comes from external services which we cannot control like external
+payment APIs, object storages, http APIs etc. It is normal to hide these kind of services behind an interface and provide mock implementation to regain
+control and determinism. However there is another source of complexity that comes from the local infrastructure that is also hard to control without building prior abstractions. Things like stdin/stdout, clocks, random generators, schedulers can make writing tests hard or even impossible. Fortunately ZIO abstracted most of it in its runtime under `Environment` type. Thanks to this design `zio-test` could easily provide its own implementation named `MockEnvironment` which gives you mocks of mentioned infrastructure. In most of the cases when you'll be using `ZIO`s `testM` mocks are already created and should be controlled by exposed functions on companion object. If for some reason you would like to provide custom environment or are using other testing framework but still want to use mocks there are `make` functions on companion objects of mocks where you can construct your own. 
+
+It is easy to accidentally use different mock instances at the same time.
+
+```scala
+testM("`acquire` doesn't leak permits upon cancellation") {
+      for {
+        mockClock <- MockClock.makeMock(MockClock.DefaultData)
+        s         <- Semaphore.make(1L)
+        sf        <- s.acquireN(2).timeout(1.milli).either.fork
+        _         <- mockClock.adjust(1.second)
+        _         <- sf.join
+        _         <- s.release
+        permits   <- s.available
+      } yield assert(permits, equalTo(2L))
+    }
+```
+
+Above code doesn't work. We created a new `MockClock` instance and are correctly adjusting its time. What might be surprising is that call to `timeout` will
+use the `MockClock` provided by the `MockEnvironment` not our `mockClock` instance. It easy to know why when you look at the signature of `timeout`:
+
+```scala
+final def timeout(d: Duration): ZIO[R with Clock, E, Option[A]] 
+```
+
+The returned type is `ZIO[R with Clock, E, Option[A]]` where our environment is "some R plus a Clock".
+Before running this `Clock` has to be provided and the framework provides the Clock from the `MockEnvironment` not our instance variable as it is not aware that we created it.
+
+
+If you need to provide real implementations instead of the mocks to some part of your tests there is a `live` method which will transform your `ZIO[R, E, A]` to `ZIO[Live[R], E, A]`. Going from `R` to `Live[R]` instructs the framework that we really want to be provided with live implementations and not mocks.
+
+
+If you have a non-trivial flow of data from multiple streams that can produce at different intervals and would like to test
+snapshots of data in particular point in time a nice pattern is to join these streams with `zipWithLatest` and 
```

Comment:
I would generalize this a little bit. `zipWithLatest` is the specific method we are testing here but it is a more general pattern. If we have a single value we're waiting for after a certain amount of time we can have the effect complete a promise, advance the clock, and wait on the promise. But if we want to advance time multiple times and wait on multiple results, using a queue gives us the same ability to "wait" for the next result without having to create multiple promises.

Reply:
OK, will talk about it and I think I will move this example to MockClock section. It makes more sense there.